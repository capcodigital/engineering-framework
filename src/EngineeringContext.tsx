import { createContext } from 'react';

const EngineeringContext = createContext({
  name: '',
  setName: (name: string) => {},
});

export default EngineeringContext;
