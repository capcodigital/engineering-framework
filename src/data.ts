export const levels = [
  'Associate',
  'Consultant',
  'Senior Consultant',
  'Principal Consultant',
  'Managing Principal',
];

export const titles = {
  'Software Engineer': {
    Associate: 'Associate Engineer',
    Consultant: 'Engineer',
    'Senior Consultant': 'Senior Engineer',
    'Principal Consultant': 'Engineering Lead / Principal Engineer',
    'Managing Principal': 'Senior Engineering Lead / Distinguished Engineer',
  },
  'Quality Engineer': {
    Associate: 'Associate Test Engineer',
    Consultant: 'Test Engineer',
    'Senior Consultant': 'Senior Test Engineer',
    'Principal Consultant': 'Test Engineering Lead / Principal Test Engineer',
    'Managing Principal':
      'Senior Test Engineering Lead / Distinguished Test Engineer',
  },
};

export const sideBarData = [
  {
    category: 'Overview',
    competencies: ['Job Description', 'Framework Criteria', 'Performance'],
  },
  {
    category: 'Delivery',
    competencies: [
      'Creating and understanding code',
      'Testing',
      'Debugging',
      'Observability',
      'Architecture',
      'Cloud',
      'Data Design & Handling',
      'Security Awareness and Best Practise Implementation',
      'Technical Leadership',
    ],
  },
  {
    category: 'People',
    competencies: [
      'Management',
      'Feedback',
      'Effective Communication',
      'Teamwork',
      'Relationship Building',
      'Inclusion',
      'Mentoring',
      'Continuous Learning',
    ],
  },
  {
    category: 'Stakeholder',
    competencies: [
      'Dealing with Ambiguity & Risk',
      'Economic Thinking',
      'Decision Making',
      'Driving Alignment',
      'Business Acumen',
      'Strategic Work',
    ],
  },
];

export const contentData = {
  Associate: {
    Overview: [],
    Delivery: [
      {
        name: 'Creating and Understanding Code',
        specialism: 'Software Engineer',
        description:
          "Is able to gain context within the team's domain; Writes code with testability, readability, edge cases, and errors in mind.",
      },
      {
        name: 'Creating and Understanding Code',
        specialism: 'Quality Engineer',
        description:
          "Able to gain context within the team's domain; Writes test code with readability, edge cases, and errors in mind.",
      },
      {
        name: 'Testing',
        specialism: 'Software Engineer',
        description:
          'Knows the testing pyramid. Writes unit tests, sometimes with help from more senior engineers.',
      },
      {
        name: 'Testing',
        specialism: 'Quality Engineer',
        description:
          'Knows the testing pyramid. Writes higher level tests, sometimes with help from more senior engineers.',
      },
      {
        name: 'Debugging',
        specialism: 'All',
        description: 'Understands the basics of debugging and the tools used. ',
      },
      {
        name: 'Observability',
        specialism: 'All',
        description: 'Not applicable at this level.',
      },
      {
        name: 'Architecture',
        specialism: 'All',
        description:
          'Is aware of overall service architecture. Designs basic functions, understands where their functions fit in the overall solution',
      },
      {
        name: 'Cloud',
        specialism: 'All',
        description:
          'Is aware of the overall landscape of cloud providers and how the organisation uses them.  Is aware of the importance of learning and understanding cloud technologies.',
      },
      {
        name: 'Data Design & Handling',
        specialism: 'All',
        description:
          'Understands the role that data plays and has worked with data in systems',
      },
      {
        name: 'Security Awereness & Best Practise Implementation',
        specialism: 'All',
        description:
          'Understands the importance and language of security concepts.',
      },
      {
        name: 'Technical Leadership',
        specialism: 'All',
        description: 'Not applicable at this level.',
      },
    ],
    People: [
      {
        name: 'Management',
        specialism: 'All',
        description: 'Not applicable at this level.',
      },
      {
        name: 'Feedback',
        specialism: 'All',
        description:
          'Understands how to deliver praise and constructive feedback in a useful manner.  Actively seeks out feedback from their teammates and manager, and works to use feedback that they receive as a tool for growth.',
      },
      {
        name: 'Effective Communication',
        specialism: 'All',
        description:
          'Communicates effectively, clearly, concisely and in an audience-oriented way, in written and verbal form.  Understands their area of work and shares their knowledge frequently with their teammates. Actively listens to others and ensures they are understood. Pays attention to nonverbal communication.',
      },
      {
        name: 'Teamwork',
        specialism: 'All',
        description:
          'Has daily conversation with the team about the progress of their work. Helps their teammates when requested. Gives or shares credit where due.',
      },
      {
        name: 'Relationship Building',
        specialism: 'All',
        description:
          'Works to build strong relationships with their teammates and manager.',
      },
      {
        name: 'Incllusion',
        specialism: 'All',
        description:
          "Openly shares their opinions and contributes to discussions in a respectful manner. Works with teammates to resolve disagreement in a healthy manner. Is open to changing their perspective and plans based on others' input. ",
      },
      {
        name: 'Mentoring',
        specialism: 'All',
        description: 'Seeks out mentorship to grow their own experience.',
      },
      {
        name: 'Continuous Learning',
        specialism: 'All',
        description:
          'Recognises the opportunities for both structured and unstructured learning and has a plan for continuous development',
      },
    ],
    Stakeholder: [
      {
        name: 'Dealing with AMbiguity & Risk',
        specialism: 'All',
        description: 'Not applicable at this level.',
      },
      {
        name: 'Economic Thinking',
        specialism: 'All',
        description:
          'Understands the importance of weighing cost and value in decision making. Asks more senior engineers for help in applying this type of thinking to their work.',
      },
      {
        name: 'Decision Making',
        specialism: 'All',
        description:
          'Can articulate decisions required, identifies relevant context and prepares options for consideration.',
      },
      {
        name: 'Driving Alignment',
        specialism: 'All',
        description: 'Not applicable at this level.',
      },
      {
        name: 'Business Acumen',
        specialism: 'All',
        description: 'Not applicable at this level.',
      },
      {
        name: 'Strategic Work',
        specialism: 'All',
        description: 'Not applicable at this level.',
      },
    ],
  },
};
