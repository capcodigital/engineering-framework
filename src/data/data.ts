export const levels = [
  "Associate",
  "Consultant",
  "Senior Consultant",
  "Principal Consultant",
  "Managing Principal",
];

export const tagText = {
  "Technical Leadership": {
    "Principal Consultant": {
      "Software Engineer": { name: "Principal Engineer", short: "PE" },
      "Quality Engineer": { name: "Principal Quality Engineer", short: "PTE" },
      color: "#E36756",
    },
    "Managing Principal": {
      "Software Engineer": { name: "Distinguished Engineer", short: "DE" },
      "Quality Engineer": {
        name: "Distinguished Quality Engineer",
        short: "DTE",
      },
      color: "#ED4664",
    },
  },
  Management: {
    "Principal Consultant": {
      "Software Engineer": { name: "Engineering Lead", short: "EL" },
      "Quality Engineer": { name: "Quality Engineering Lead", short: "QEL" },
      color: "#5686E3",
    },
    "Managing Principal": {
      "Software Engineer": { name: "Senior Engineering Lead", short: "SEL" },
      "Quality Engineer": {
        name: "Senior Quality Engineering Lead",
        short: "SQEL",
      },
      color: "#8356E3",
    },
  },
};

export const titles = {
  "Software Engineer": {
    Associate: "Associate Engineer",
    Consultant: "Engineer",
    "Senior Consultant": "Senior Engineer",
    "Principal Consultant": "Engineering Lead / Principal Engineer",
    "Managing Principal": "Senior Engineering Lead / Distinguished Engineer",
  },
  "Quality Engineer": {
    Associate: "Associate Quality Engineer",
    Consultant: "Quality Engineer",
    "Senior Consultant": "Senior Quality Engineer",
    "Principal Consultant":
      "Quality Engineering Lead / Principal Quality Engineer",
    "Managing Principal":
      "Senior Quality Engineering Lead / Distinguished Quality Engineer",
  },
};

export const sideBarData = [
  {
    category: "Overview",
    competencies: ["Description", "You Have", "You Will"],
  },
  {
    category: "Delivery",
    competencies: [
      "Narrative",
      "Technical Leadership",
      "Coding",
      "Testing",
      "Debugging",
      "Observability",
      "Architecture",
      "Cloud",
      "Data Design & Handling",
      "Security",
    ],
  },
  {
    category: "People",
    competencies: [
      "Narrative",
      "Management",
      "Feedback",
      "Effective Communication",
      "Teamwork",
      "Relationship Building",
      "Inclusion",
      "Mentoring",
      "Continuous Learning",
    ],
  },
  {
    category: "Stakeholder",
    competencies: [
      "Narrative",
      "Ambiguity & Risk",
      "Economic Thinking",
      "Decision Making",
      "Driving Alignment",
      "Business Acumen",
      "Strategic Work",
    ],
  },
];
