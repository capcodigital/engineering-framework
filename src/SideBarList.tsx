import React, { useContext } from 'react';
import { NavHashLink as Link } from 'react-router-hash-link';
import { EngineeringContext } from './EngineeringContext';
import { Collapse, List, ListItem } from '@material-ui/core';
import styled from 'styled-components/macro';
import { SmallTag } from './Tag';
import { sideBarData, levels } from './data/data';

const StyledList = styled(List)`
  && {
    min-width: 300px;
    color: grey;
    a {
      text-decoration: none;
    }
    .active {
      color: white;
    }
    .nested {
      margin-left: 15px;
    }
    .list-item {
      height: 32px;
      margin-bottom: 8px;
    }
    .level.list-item {
      font-size: 20px;
      font-weight: bold;
    }
    .category.list-item {
      font-size: 16px;
      font-weight: bold;
      a {
        color: grey;
      }
    }
    .category.list-item.active {
      border-left: 3px solid #e72871;
      a {
        color: white;
      }
    }
    .competency.list-item {
      font-size: 16px;
      a {
        color: grey;
      }
    }
    .competency.list-item.active {
      a {
        color: white;
      }
    }
  }
`;

const SideBarList = () => {
  const {
    level,
    category,
    competency,
    specialism,
    setLevel,
    setCategory,
    setCompetency,
  } = useContext(EngineeringContext);

  const handleClickCategory = (
    categoryName: string,
    firstCompetency: string
  ) => {
    setCategory(categoryName);
    setCompetency(firstCompetency);
  };
  const handleClickCompetency = (competencyName: string) => {
    setCompetency(competencyName);
  };
  const handleClickLevel = (levelName: string) => {
    setLevel(levelName);
  };

  return (
    <div>
      {levels.map((levelName: string) => (
        <StyledList className='list' key={levelName}>
          <ListItem
            button
            className={`level list-item ${levelName === level ? 'active' : ''}`}
            onClick={() => handleClickLevel(levelName)}
          >
            {levelName}
          </ListItem>
          <Collapse in={level === levelName} timeout='auto' unmountOnExit>
            <List component='nav' className={'nested'}>
              {sideBarData.map((data: any) => (
                <span key={data.category}>
                  <ListItem
                    button
                    data-testid={`${levelName}-${data.category}`.toLowerCase()}
                    className={`category list-item ${
                      category === data.category ? 'active' : ''
                    }`}
                    onClick={() =>
                      handleClickCategory(data.category, data.competencies[0])
                    }
                  >
                    <Link
                      smooth
                      to={`#${data.category.toLowerCase()}-${data.competencies[0]
                        .replace(/\s/g, '-')
                        .toLowerCase()}`}
                    >
                      {data.category}
                    </Link>
                  </ListItem>
                  <Collapse
                    in={category === data.category}
                    timeout='auto'
                    unmountOnExit
                  >
                    <List component='div' disablePadding>
                      {data.competencies.map((competencyName: string) => {
                        // Don't have Framework Criteria at MP level
                        if (
                          level !== 'Managing Principal' ||
                          competencyName !== 'Framework Criteria'
                        )
                          return (
                            <ListItem
                              key={competencyName}
                              button
                              className={`nested competency list-item ${
                                competency === competencyName ? 'active' : ''
                              } `}
                              onClick={() =>
                                handleClickCompetency(competencyName)
                              }
                            >
                              <Link
                                smooth
                                to={`#${data.category}-${competencyName.replace(
                                  /\s/g,
                                  '-'
                                )}`.toLowerCase()}
                              >
                                {competencyName}

                                <SmallTag
                                  level={level}
                                  competency={competencyName}
                                  specialism={specialism}
                                />
                              </Link>
                            </ListItem>
                          );
                      })}
                    </List>
                  </Collapse>
                </span>
              ))}
            </List>
          </Collapse>
        </StyledList>
      ))}
    </div>
  );
};

export default SideBarList;
