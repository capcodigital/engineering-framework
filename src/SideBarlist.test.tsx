import * as React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import SideBarList from './SideBarList';
import { levels, sideBarData } from './data/data';
import { BrowserRouter as Router } from 'react-router-dom';
import { EngineeringContextProvider } from './EngineeringContext';

describe('SideBarList', () => {
  beforeEach(() => {
    render(
      <Router>
        <EngineeringContextProvider>
          <SideBarList />
        </EngineeringContextProvider>
      </Router>
    );
  });
  
  it('should render SideBarList component', () => {
    const { container } = render(
      <Router>
        <EngineeringContextProvider>
          <SideBarList />
        </EngineeringContextProvider>
      </Router>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  levels.map((levelText: string) => {
    it(`should check that all ${levelText} is visible in the side bar list`, () => {
      expect(screen.getByText(levelText)).toBeVisible();
    });
  });

  describe('Default behaviour on first load', () => {
    it('should check that Associate level is active', () => {
      expect(screen.getByText('Associate')).toHaveClass('active');
      expect(screen.getByText('Associate')).toHaveStyle('color:white');
    });

    it('should check that other than Associate level, all levels are not active', () => {
      levels.slice(1).map((leveltext: string) => {
        expect(screen.getByText(leveltext)).not.toHaveClass('active');
        expect(screen.getByText(leveltext)).not.toHaveStyle('color:white');
      });
    });

    it('should check that all categories are visible', () => {
      sideBarData
        .map((data: any) => data.category)
        .map((categoryName: string) => {
          expect(screen.getByText(categoryName)).toBeVisible();
        });
    });

    it('should check that Overview is active', () => {
      expect(screen.getByTestId('associate-overview')).toHaveClass('active');
      expect(screen.getByTestId('associate-overview')).toHaveStyle(
        'color:white'
      );
    });

    it('should check that other than Overview, all categories are inactive', () => {
      sideBarData
        .slice(1)
        .map((data: any) => data.category)
        .map((categoryName: string) => {
          expect(
            screen.getByTestId(`associate-${categoryName.toLowerCase()}`)
          ).not.toHaveStyle('color:white');
          expect(
            screen.getByTestId(`associate-${categoryName.toLowerCase()}`)
          ).not.toHaveClass('active');
        });
    });

    it('should check that all competencies are visible under Overview', () => {
      sideBarData
        .find((data: any) => {
          data.category === 'Overview';
        })
        ?.competencies.map((competencyName: string) => {
          expect(screen.getByText(competencyName)).toBeVisible();
        });
    });

    it('should check that Job Description is active', () => {
      expect(screen.getByText('Job Description'!)).toHaveStyle('color:white');
    });

    it('should check that other than Job Description, all competencies are inactive', () => {
      sideBarData
        .slice(1)
        .find((data: any) => data.category === 'Overview')
        ?.competencies.map((competencyName: string) => {
          expect(screen.getByText(competencyName)).not.toHaveStyle(
            'color:white'
          );
        });
    });
  });

  describe('Select different categories', () => {
    it('should open Delivery and check that it is active ie text is white', () => {
      fireEvent.click(screen.getByTestId('associate-delivery'));
      expect(screen.getByTestId('associate-delivery')).toHaveClass('active');
      expect(screen.getByTestId('associate-delivery')).toHaveStyle(
        'color:white'
      );
    });
    it('should click Delivery and check all competencies are visible', () => {
      fireEvent.click(screen.getByTestId('associate-delivery'));
      sideBarData
        .find((data: any) => {
          data.category === 'Delivery';
        })
        ?.competencies.map((competencyName: string) => {
          expect(screen.getByText(competencyName)).toBeVisible();
        });
    });

    it('should click Delivery and check that first competency is active ie text is white', () => {
      fireEvent.click(screen.getByTestId('associate-delivery'));
      let firstCompetency = sideBarData.find(
        (data: any) => data.category === 'Delivery'
      )?.competencies[0];

      expect(screen.getByText(firstCompetency!)).toBeVisible();
      expect(screen.getByText(firstCompetency!)).toHaveStyle('color:white');
    });

    it('should click People and check all competencies are visible', () => {
      fireEvent.click(screen.getByTestId('associate-people'));
      sideBarData
        .find((data: any) => data.category === 'People')
        ?.competencies.map((competencyName: string) =>
          expect(screen.getByText(competencyName)).toBeVisible()
        );
    });

    it('should click Stakeholder and check all competencies are visible', () => {
      fireEvent.click(screen.getByTestId('associate-stakeholder'));
      sideBarData
        .find((data: any) => data.category === 'Stakeholder')
        ?.competencies.map((competencyName: string) =>
          expect(screen.getByText(competencyName)).toBeVisible()
        );
    });
  });
});
