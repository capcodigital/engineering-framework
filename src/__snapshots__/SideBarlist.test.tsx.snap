// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBarList should render SideBarList component 1`] = `
<div>
  <ul
    class="MuiList-root SideBarList__StyledList-sc-1egehub-0 gLBXYT list MuiList-padding"
  >
    <div
      aria-disabled="false"
      class="MuiButtonBase-root MuiListItem-root level list-item  MuiListItem-gutters MuiListItem-button"
      role="button"
      tabindex="0"
    >
      Associate
      <span
        class="MuiTouchRipple-root"
      />
    </div>
  </ul>
  <ul
    class="MuiList-root SideBarList__StyledList-sc-1egehub-0 gLBXYT list MuiList-padding"
  >
    <div
      aria-disabled="false"
      class="MuiButtonBase-root MuiListItem-root level list-item  MuiListItem-gutters MuiListItem-button"
      role="button"
      tabindex="0"
    >
      Consultant
      <span
        class="MuiTouchRipple-root"
      />
    </div>
  </ul>
  <ul
    class="MuiList-root SideBarList__StyledList-sc-1egehub-0 gLBXYT list MuiList-padding"
  >
    <div
      aria-disabled="false"
      class="MuiButtonBase-root MuiListItem-root level list-item  MuiListItem-gutters MuiListItem-button"
      role="button"
      tabindex="0"
    >
      Senior Consultant
      <span
        class="MuiTouchRipple-root"
      />
    </div>
  </ul>
  <ul
    class="MuiList-root SideBarList__StyledList-sc-1egehub-0 gLBXYT list MuiList-padding"
  >
    <div
      aria-disabled="false"
      class="MuiButtonBase-root MuiListItem-root level list-item  MuiListItem-gutters MuiListItem-button"
      role="button"
      tabindex="0"
    >
      Principal Consultant
      <span
        class="MuiTouchRipple-root"
      />
    </div>
  </ul>
  <ul
    class="MuiList-root SideBarList__StyledList-sc-1egehub-0 gLBXYT list MuiList-padding"
  >
    <div
      aria-disabled="false"
      class="MuiButtonBase-root MuiListItem-root level list-item  MuiListItem-gutters MuiListItem-button"
      role="button"
      tabindex="0"
    >
      Managing Principal
      <span
        class="MuiTouchRipple-root"
      />
    </div>
  </ul>
</div>
`;
